rustfmt.toml

```diff
@@ -0,0 +1,3 @@
+imports_granularity="Crate"
+group_imports="StdExternalCrate"
+reorder_imports=true
```

Comment:
There are multiple Cargo projects or workspace.

```toml
# file: Cargo.toml

exclude = [
    "osdk",
    "services/libs/comp-sys/cargo-component",
    "services/libs/comp-sys/component",
    "services/libs/comp-sys/component-macro",
    "services/libs/comp-sys/controlled",
]
```

None of the Cargo projects under the `exclude` paths have been updated. So it seems that the rustfmt tool or the new format rules does not affect these projects. Do we need to add `rustfmt.toml` to each of the excluded Rust projects or what?

Reply:
I've figured out the reason why this is happening. The `cargo fmt` command only affects crates inside the workspace, which means running `cargo fmt` in the root directory doesn't impact the excluded crates, even if a rustfmt.toml is added. 

To format these crates, you need to move to the corresponding crate directories and execute `cargo fmt` there. In this case, adding an extra rustfmt.toml is unnecessary because it will automatically parse the rustfmt.toml from the upper directories. 

I've added a additional script named `format_all.sh` in the root directory. Running this script will automatically handle the directory movement and the `cargo fmt` operation. Alternatively, developers can run cargo fmt themselves while working in the respective directories. 

However, this issue has also led me to realize that our current `make check` only checks the crates within the workspace, which means our testing scope is not comprehensive enough.

Reply:
Now that we have realized that there are more crates to be formatted, I think that the root Makefile should be extended.

```rust
format:
    ./tools/format_all.sh

check:
    ./tools/format_all.sh --check
```